File: java/src/main/java/com/dokmatiq/docgen/DocGen.java
package com.dokmatiq.docgen;

import com.dokmatiq.docgen.builder.ComposeBuilder;
import com.dokmatiq.docgen.builder.DocumentBuilder;
import com.dokmatiq.docgen.builder.InvoiceBuilder;
import com.dokmatiq.docgen.client.*;
import com.dokmatiq.docgen.internal.FileUtils;
import com.dokmatiq.docgen.internal.Transport;
import com.dokmatiq.docgen.model.OutputFormat;

import java.nio.file.Path;
import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;

/**
 * Main DocGen SDK client.
 *
 * <pre>{@code
 * try (var dg = new DocGen(DocGenConfig.builder("dk_live_xxx").build())) {
 *     // One-liner
 *     byte[] pdf = dg.htmlToPdf("<h1>Hello World</h1>");
 *
 *     // Builder
 *     byte[] invoice = dg.document()
 *         .html("<h1>Rechnung</h1>")
 *         .template("invoice.odt")
 *         .field("nr", "2026-001")
 *         .asPdf()
 *         .generate();
 * }
 * }</pre>
 */
public class DocGen implements AutoCloseable {
    private final Transport transport;

    private final DocumentsClient documents;
    private final TemplatesClient templates;
    private final FontsClient fonts;
    private final PdfFormsClient pdfForms;
    private final SignaturesClient signatures;
    private final PdfToolsClient pdfTools;
    private final PreviewClient preview;
    private final ZugferdClient zugferd;
    private final XRechnungClient xrechnung;
    private final ExcelClient excel;
    private final ReceiptsClient receipts;

    public DocGen(DocGenConfig config) {
        this.transport = new Transport(config);
        this.documents = new DocumentsClient(transport);
        this.templates = new TemplatesClient(transport);
        this.fonts = new FontsClient(transport);
        this.pdfForms = new PdfFormsClient(transport);
        this.signatures = new SignaturesClient(transport);
        this.pdfTools = new PdfToolsClient(transport);
        this.preview = new PreviewClient(transport);
        this.zugferd = new ZugferdClient(transport);
        this.xrechnung = new XRechnungClient(transport);
        this.excel = new ExcelClient(transport);
        this.receipts = new ReceiptsClient(transport);
    }

    /** Convenience: create with just an API key. */
    public DocGen(String apiKey) {
        this(DocGenConfig.builder(apiKey).build());
    }

    // ── Sub-clients ─────────────────────────────────────────────────

    public DocumentsClient documents() { return documents; }
    public TemplatesClient templates() { return templates; }
    public FontsClient fonts() { return fonts; }
    public PdfFormsClient pdfForms() { return pdfForms; }
    public SignaturesClient signatures() { return signatures; }
    public PdfToolsClient pdfTools() { return pdfTools; }
    public PreviewClient preview() { return preview; }
    public ZugferdClient zugferd() { return zugferd; }
    public XRechnungClient xrechnung() { return xrechnung; }
    public ExcelClient excel() { return excel; }
    public ReceiptsClient receipts() { return receipts; }

    // ── Builder entry points ────────────────────────────────────────

    /** Create a DocumentBuilder for fluent document construction. */
    public DocumentBuilder document() {
        return new DocumentBuilder().setClient(documents);
    }

    /** Create a ComposeBuilder for multi-part composition. */
    public ComposeBuilder compose() {
        return new ComposeBuilder().setClient(documents);
    }

    /** Create an InvoiceBuilder for structured invoice data. */
    public InvoiceBuilder invoice() {
        return new InvoiceBuilder();
    }

    // ── Convenience methods ─────────────────────────────────────────

    /** Convert HTML to PDF in one call. */
    public byte[] htmlToPdf(String html) {
        return document().html(html).asPdf().generate();
    }

    /** Convert Markdown to PDF in one call. */
    public byte[] markdownToPdf(String markdown) {
        return document().markdown(markdown).asPdf().generate();
    }

    /** Merge multiple PDF files into one. */
    public byte[] mergePdfs(List<Path> files) {
        return pdfTools.merge(files);
    }

    /** Sign a PDF with a certificate. */
    public byte[] signPdf(Path file, String certificateName, String certificatePassword) {
        return signatures.sign(file, certificateName, certificatePassword);
    }

    /** Fill form fields in a PDF. */
    public byte[] fillForm(Path file, Map<String, String> fields) {
        return pdfForms.fill(file, fields, false);
    }

    /** Fill form fields in a PDF with flatten option. */
    public byte[] fillForm(Path file, Map<String, String> fields, boolean flatten) {
        return pdfForms.fill(file, fields, flatten);
    }

    @Override
    public void close() {
        transport.close();
    }
}

File: php/src/DocGen.php
<?php

declare(strict_types=1);

namespace Dokmatiq\DocGen;

use Dokmatiq\DocGen\Builder\ComposeBuilder;
use Dokmatiq\DocGen\Builder\DocumentBuilder;
use Dokmatiq\DocGen\Builder\InvoiceBuilder;
use Dokmatiq\DocGen\Client\DocumentsClient;
use Dokmatiq\DocGen\Client\FontsClient;
use Dokmatiq\DocGen\Client\PdfFormsClient;
use Dokmatiq\DocGen\Client\PdfToolsClient;
use Dokmatiq\DocGen\Client\PreviewClient;
use Dokmatiq\DocGen\Client\SignaturesClient;
use Dokmatiq\DocGen\Client\TemplatesClient;
use Dokmatiq\DocGen\Client\ExcelClient;
use Dokmatiq\DocGen\Client\ReceiptsClient;
use Dokmatiq\DocGen\Client\XRechnungClient;
use Dokmatiq\DocGen\Client\ZugferdClient;
use Dokmatiq\DocGen\Internal\FileUtils;
use Dokmatiq\DocGen\Internal\Transport;
use Dokmatiq\DocGen\Model\DocumentRequest;
use Dokmatiq\DocGen\Model\OutputFormat;

/**
 * DocGen PHP SDK – main client.
 *
 * Usage:
 *   $dg = new DocGen('your-api-key');
 *
 *   // Fluent builder
 *   $pdf = $dg->document()
 *       ->html('<h1>Hello {{name}}</h1>')
 *       ->field('name', 'World')
 *       ->asPdf()
 *       ->generate();
 *
 *   // Convenience
 *   $pdf = $dg->htmlToPdf('<h1>Hello</h1>');
 *
 *   // Sub-clients
 *   $dg->templates()->upload('invoice.odt');
 *   $dg->pdfTools()->merge(['a.pdf', 'b.pdf']);
 */
final class DocGen
{
    private readonly Transport $transport;

    public readonly DocumentsClient $documents;
    public readonly TemplatesClient $templates;
    public readonly FontsClient $fonts;
    public readonly PdfFormsClient $pdfForms;
    public readonly SignaturesClient $signatures;
    public readonly PdfToolsClient $pdfTools;
    public readonly PreviewClient $preview;
    public readonly ZugferdClient $zugferd;
    public readonly XRechnungClient $xrechnung;
    public readonly ExcelClient $excel;
    public readonly ReceiptsClient $receipts;

    public function __construct(
        string $apiKey,
        string $baseUrl = 'https://api.dokmatiq.com',
        int $timeout = 120,
        int $maxRetries = 3,
    ) {
        $config = new DocGenConfig($apiKey, $baseUrl, $timeout, $maxRetries);
        $this->transport = new Transport($config);

        $this->documents  = new DocumentsClient($this->transport);
        $this->templates  = new TemplatesClient($this->transport);
        $this->fonts      = new FontsClient($this->transport);
        $this->pdfForms   = new PdfFormsClient($this->transport);
        $this->signatures = new SignaturesClient($this->transport);
        $this->pdfTools   = new PdfToolsClient($this->transport);
        $this->preview    = new PreviewClient($this->transport);
        $this->zugferd    = new ZugferdClient($this->transport);
        $this->xrechnung  = new XRechnungClient($this->transport);
        $this->excel      = new ExcelClient($this->transport);
        $this->receipts   = new ReceiptsClient($this->transport);
    }

    /** Alternative constructor from config object. */
    public static function fromConfig(DocGenConfig $config): self
    {
        $instance = new self($config->apiKey, $config->baseUrl, $config->timeout, $config->maxRetries);
        return $instance;
    }

    // ── Fluent Builders ─────────────────────────────────────────────

    /** Start building a document request. */
    public function document(): DocumentBuilder
    {
        return new DocumentBuilder($this->documents);
    }

    /** Start building a compose (multi-part) request. */
    public function compose(): ComposeBuilder
    {
        return new ComposeBuilder($this->documents);
    }

    /** Start building invoice data. */
    public function invoice(): InvoiceBuilder
    {
        return new InvoiceBuilder();
    }

    // ── Convenience Methods ─────────────────────────────────────────

    /** Convert HTML to PDF. */
    public function htmlToPdf(string $html): string
    {
        return $this->document()->html($html)->asPdf()->generate();
    }

    /** Convert Markdown to PDF. */
    public function markdownToPdf(string $markdown): string
    {
        return $this->document()->markdown($markdown)->asPdf()->generate();
    }

    /**
     * Merge multiple PDFs.
     *
     * @param string[] $pdfPaths
     */
    public function mergePdfs(array $pdfPaths): string
    {
        return $this->pdfTools->merge($pdfPaths);
    }

    /** Sign a PDF with a certificate. */
    public function signPdf(string $pdfPath, string $certAlias, ?string $reason = null): string
    {
        return $this->signatures->sign($pdfPath, $certAlias, $reason);
    }
}
